### PR TITLE
feat: update generic swift clients with endpoints resolver

### PR DESCRIPTION
**Pull request by AndrewFossAWS**
## Description of changes
* Update generic swift clients with endpoints (See `EndpointResolverMiddleware.kt`, `EndpointResolverGenerator.kt`, `EndpointParamsGenerator.kt`)
* Moved most of the endpoints related swift classes from aws-swift-sdk to here
* Verified endpoints related classes (middleware, resolver, etc) are generated for the WeatherSDK

The following files are moved from aws-sdk-swift. 
```
Sources/ClientRuntime/Endpoints/CredentialScope.swift
Sources/ClientRuntime/Endpoints/EndpointError.swift
Sources/ClientRuntime/Endpoints/EndpointsRuleEngine.swift
Sources/ClientRuntime/Endpoints/PartitionDefinition.swift
Sources/ClientRuntime/Endpoints/ServiceEndpointMetadata.swift
Sources/ClientRuntime/Signing/AWSSigningAlgorithm.swift
```
The following are also moved from aws-sdk-swift and renamed with `Endpoints` as prefix

```
Sources/ClientRuntime/Endpoints/EndpointsAuthSchemeResolver.swift
Sources/ClientRuntime/Endpoints/EndpointsRequestContext.swift
```
The following are also moved from aws-sdk-swift and renamed with `CRT` as prefix
```
Sources/ClientRuntime/Endpoints/CRTResolvedEndpoint.swift
Sources/ClientRuntime/Endpoints/CRTsResolvedEndpointType.swift
```

CRT's EndpointsRuleEngine from aws-common-runtime requires a partition JSON string passed into its Initializer.
Partition is also an AWS concept. For smithy-swift/generic clients, we will simply pass in an empty json for the rule engine to still work i.e.
```
{
  "version": "1.1",
  "partitions": []
}
```
See `Tests/ClientRuntimeTests/Endpoints/EndpointsRuleEngineTests.swift`

Changes to WeatherSDK and WeatherSDKTests will be included in subsequent PR.

## Scope

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/WeatherSDK/EndpointResolver.swift [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
...\"stringEquals\",\"argv\":[{\"ref\":\"Stage\"},\"alpha\"]}],\"endpoint\":{\"url\":\"https://fpmhmw9kd7.execute-api.us-west-2.amazonaws.com/beta\",\"properties\":{},\"headers\":{}},\"type\":\"endpoint\"},{\"conditions\":[{\"fn\":\"isSet\",\"argv\":[{\"ref\":\"Stage\"}]},{\"fn\":\"stringEquals\",\"argv\":[{\"ref\":\"Stage\"},\"beta\"]}],\"endpoint\":{\"url\":\"https://fpmhmw9kd7.execute-api.us-west-2.amazonaws.com/beta\",\"properties\":{},\"headers\":{}},\"type\":\"endpoint\"},{\"conditions\":[{\"fn\":\"isSet\",\"argv\":[{\"ref\":\"Stage\"}]},{\"fn\":\"stringEquals\",\"argv\":[{\"ref\":\"Stage\"},\"gamma\"]}],\"endpoint\":{\"url\":\"https://mow32it2pa.execute-api.us-east-1.amazonaws.com/gamma\",\"properties\":{},\"headers\":{}},\"type\":\"endpoint\"},{\"conditions\":[{\"fn\":\"isSet\",\"argv\":[{\"ref\":\"Stage\"}]},{\"fn\":\"stringEquals\",\"argv\":[{\"ref\":\"Stage\"},\"prod\"]}],\"endpoint\":{\"url\":\"https://bqcgz0nn7a.execute-api.us-east-1.amazonaws.com/prod\",\"properties\":{},\"headers\":{}},\"type\":\"endpoint\"},{\"conditions\":[],\"endpoint\":{\"url\":\"https://bqcgz0nn7a.execute-api.us-east-1.amazonaws.com/prod\",\"properties\":{},\"headers\":{}},\"type\":\"endpoint\"}]}"
```
Comment (jbelkins) [inline Sources/WeatherSDK/EndpointResolver.swift:24]: What exactly are the hosts pointed to in these rules?

File: Sources/ClientRuntime/Endpoints/EndpointsAuthSchemeResolver.swift [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    case sigV4(SigV4Parameters)
+    case sigV4A(SigV4AParameters)
+    case none
+
+    /// The name of the auth scheme
+    public var name: String {
+        switch self {
+        case .sigV4: return "sigv4"
+        case .sigV4A: return "sigv4a"
+        case .none: return "none"
```
Comment (jbelkins) [inline Sources/ClientRuntime/Endpoints/EndpointsAuthSchemeResolver.swift:19]: Should smithy-swift include references to either sigv4 or v4a?

**Review by jbelkins — COMMENTED** [review]
A couple questions in the comments.

Also, since many of the "new" files in this PR were copied from SDK and it's not clear what changed in the transition, can you please annotate what changed in the files with significant changes (i.e. more than just adjusting package, imports, etc.)

**Review by syall — APPROVED** [review]

File: Sources/WeatherSDK/EndpointResolver.swift (hunk 1 above)
Comment (AndrewFossAWS) [reply Sources/WeatherSDK/EndpointResolver.swift:24]: I was just verifying the endpoint rulesSet gets generated correctly using fake urls. I will remove this.

File: Tests/ClientRuntimeTests/Endpoints/EndpointsRuleEngineTests.swift [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
..."Stage\":{\"required\":false,\"type\":\"String\",\"documentation\":\"docs\"}},\"rules\":[{\"type\":\"endpoint\",\"conditions\":[{\"fn\":\"isSet\",\"argv\":[{\"ref\":\"Stage\"}]},{\"fn\":\"stringEquals\",\"argv\":[{\"ref\":\"Stage\"},\"alpha\"]}],\"endpoint\":{\"url\":\"https://fpmhmw9kd7.execute-api.us-west-2.amazonaws.com/beta\"}},{\"type\":\"endpoint\",\"conditions\":[{\"fn\":\"isSet\",\"argv\":[{\"ref\":\"Stage\"}]},{\"fn\":\"stringEquals\",\"argv\":[{\"ref\":\"Stage\"},\"beta\"]}],\"endpoint\":{\"url\":\"https://fpmhmw9kd7.execute-api.us-west-2.amazonaws.com/beta\"}},{\"type\":\"endpoint\",\"conditions\":[{\"fn\":\"isSet\",\"argv\":[{\"ref\":\"Stage\"}]},{\"fn\":\"stringEquals\",\"argv\":[{\"ref\":\"Stage\"},\"gamma\"]}],\"endpoint\":{\"url\":\"https://mow32it2pa.execute-api.us-east-1.amazonaws.com/gamma\"}},{\"type\":\"endpoint\",\"conditions\":[{\"fn\":\"isSet\",\"argv\":[{\"ref\":\"Stage\"}]},{\"fn\":\"stringEquals\",\"argv\":[{\"ref\":\"Stage\"},\"prod\"]}],\"endpoint\":{\"url\":\"https://bqcgz0nn7a.execute-api.us-east-1.amazonaws.com/prod\"}},{\"type\":\"endpoint\",\"conditions\":[],\"endpoint\":{\"url\":\"https://bqcgz0nn7a.execute-api.us-east-1.amazonaws.com/prod\"}}]}")
```
[the remark this answers is not in the document]
Comment (AndrewFossAWS) [reply Tests/ClientRuntimeTests/Endpoints/EndpointsRuleEngineTests.swift]: updated with a test endpoint rules json file

**Pull request merged**